festim/exports/txt_export.py

```diff
@@ -51,83 +67,101 @@ def is_it_time_to_export(self, current_time):
                 return True
         return False
 
-    def when_is_next_time(self, current_time):
-        if self.times is None:
-            return None
-        for time in self.times:
-            if current_time < time:
-                return time
-        return None
+    def is_last(self, current_time, final_time):
+        if final_time is None:
+            # write if steady
+            return True
+        elif self.times is None:
+            if np.isclose(current_time, final_time, atol=0):
+                # write at final time if exports at each timestep
+                return True
+        else:
+            if np.isclose(current_time, self.times[-1], atol=0):
+                # write at final time if exports at specific times
+                return True
+        return False
+
+    def initialise_TXTExport(self, mesh, project_to_DG=False, materials=None):
 
-    def write(self, current_time, steady):
-        # create a DG1 functionspace
-        V_DG1 = f.FunctionSpace(self.function.function_space().mesh(), "DG", 1)
+        if project_to_DG:
+            self._V = f.FunctionSpace(mesh, "DG", 1)
+        else:
+            self._V = f.FunctionSpace(mesh, "CG", 1)
+
+        x = f.interpolate(f.Expression("x[0]", degree=1), self._V)
+        x_column = np.transpose([x.vector()[:]])
+
+        # if chemical_pot is True or trap_element_type is DG, get indices of duplicates near interfaces
+        # and indices of the first elements from a pair of duplicates otherwise
+        if project_to_DG:
+            # Collect all borders
+            borders = []
+            for material in materials:
+                if material.borders:
+                    for border in material.borders:
+                        borders.append(border)
+            borders = np.unique(borders)
+
+            # Find indices of the closest duplicates to interfaces
+            border_indices = []
+            for border in borders:
+                closest_indx = np.abs(x_column - border).argmin()
+                closest_x = x_column[closest_indx]
+                for ind in np.where(x_column == closest_x)[0]:
+                    border_indices.append(ind)
+
+            # Find indices of first elements in duplicated pairs and mesh borders
+            _, mesh_indices = np.unique(x_column, return_index=True)
+
+            # Get unique indices from both arrays preserving the order in unsorted x-array
+            unique_indices = []
+            for indx in np.argsort(x_column, axis=0)[:, 0]:
+                if (indx in mesh_indices) or (indx in border_indices):
+                    unique_indices.append(indx)
+
+            self._unique_indices = np.array(unique_indices)
+
+        else:
+            # Get list of unique indices as integers
+            self._unique_indices = np.argsort(x_column, axis=0)[:, 0]
+
+        self.data = x_column[self._unique_indices]
+        self.header = "x"
```

Comment:
This method now initialises the exported data. `self._unique_indices` allows to get a sorted array without duplicates (or with those near interfaces) from an unsorted array of initial data. Possibly, there is a more efficient way of doing it?